File: camp-card-android/app/src/main/java/org/bsa/campcard/core/auth/AuthViewModel.kt
package org.bsa.campcard.core.auth

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import org.bsa.campcard.BuildConfig
import org.bsa.campcard.core.models.User
import org.bsa.campcard.core.models.RegisterRequest
import javax.inject.Inject

data class AuthState(
    val isLoading: Boolean = false,
    val isAuthenticated: Boolean = false,
    val user: User? = null,
    val error: String? = null
)

@HiltViewModel
class AuthViewModel @Inject constructor(
    private val authRepository: AuthRepository
) : ViewModel() {

    private val _state = MutableStateFlow(AuthState())
    val state: StateFlow<AuthState> = _state.asStateFlow()

    init {
        initialize()
    }

    private fun initialize() {
        if (BuildConfig.DEBUG && !authRepository.isLoggedIn()) {
            // Debug auto-login with demo account (mirrors iOS screenshot mode)
            viewModelScope.launch {
                try {
                    val resp = authRepository.login("demo@campcard.org", "Password123")
                    _state.update { it.copy(isLoading = false, isAuthenticated = true, user = resp.user) }
                } catch (_: Exception) {}
            }
            return
        }
        if (!authRepository.isLoggedIn()) return
        _state.update { it.copy(isLoading = true) }
        viewModelScope.launch {
            try {
                val user = authRepository.me()
                _state.update { it.copy(isLoading = false, isAuthenticated = true, user = user) }
            } catch (e: Exception) {
                authRepository.logout()
                _state.update { AuthState() }
            }
        }
    }

    fun login(email: String, password: String) {
        _state.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            try {
                val resp = authRepository.login(email, password)
                authRepository.storeBiometricCredentials(email, resp.refreshToken)
                _state.update { it.copy(isLoading = false, isAuthenticated = true, user = resp.user) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "Login failed") }
            }
        }
    }

    fun register(request: RegisterRequest) {
        _state.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            try {
                val resp = authRepository.register(request)
                _state.update { it.copy(isLoading = false, isAuthenticated = true, user = resp.user) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = e.message ?: "Registration failed") }
            }
        }
    }

    fun loginWithBiometric(refreshToken: String) {
        _state.update { it.copy(isLoading = true, error = null) }
        viewModelScope.launch {
            try {
                val resp = authRepository.refresh(refreshToken)
                _state.update { it.copy(isLoading = false, isAuthenticated = true, user = resp.user) }
            } catch (e: Exception) {
                _state.update { it.copy(isLoading = false, error = "Biometric login failed") }
            }
        }
    }

    fun logout() {
        viewModelScope.launch {
            authRepository.logout()
            _state.update { AuthState() }
        }
    }

    fun updateUser(user: User) {
        _state.update { it.copy(user = user) }
    }

    fun clearError() {
        _state.update { it.copy(error = null) }
    }

    val biometricRefreshToken: String? get() = authRepository.biometricRefreshToken
    val hasBiometricCredentials: Boolean get() = authRepository.biometricRefreshToken != null
}

File: camp-card-ios/Sources/Core/Auth/AuthViewModel.swift
import Foundation
import SwiftUI
import LocalAuthentication

@MainActor
final class AuthViewModel: ObservableObject {
    static let shared = AuthViewModel()

    @Published var user: User?
    @Published var isAuthenticated = false
    @Published var isInitializing = true   // true only during cold-start token check
    @Published var isLoading = false
    @Published var error: String?

    private let api = APIClient.shared
    private let keychain = KeychainService.shared

    private init() {
        Task { await initialize() }
    }

    // MARK: - Initialize (on app launch)
    func initialize() async {
        defer { isInitializing = false }
        // Screenshot mode: auto-login with demo account so real API data loads
        if ProcessInfo.processInfo.environment["SCREENSHOT_MODE"] == "1" {
            try? await login(email: "demo@campcard.org", password: "Password123")
            return
        }
        guard keychain.accessToken != nil else { return }
        do {
            let u: User = try await api.request(.me)
            user = u
            isAuthenticated = true
        } catch {
            keychain.clearAll()
        }
    }

    // MARK: - Login
    func login(email: String, password: String) async throws {
        isLoading = true
        defer { isLoading = false }
        error = nil
        do {
            let resp: AuthResponse = try await api.request(.login(email: email, password: password))
            keychain.storeAuthTokens(access: resp.accessToken, refresh: resp.refreshToken)
            keychain.storeBiometricCredentials(email: email, refreshToken: resp.refreshToken)
            user = resp.user
            isAuthenticated = true
        } catch let ne as NetworkError {
            error = ne.errorDescription
            throw ne
        }
    }

    // MARK: - Biometric Login
    var canUseBiometric: Bool {
        let context = LAContext()
        var error: NSError?
        let canEval = context.canEvaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, error: &error)
        if canEval { return true }
        // Show button if hardware supports biometrics but isn't enrolled yet
        if let laError = error as? LAError, laError.code == .biometryNotEnrolled { return true }
        return false
    }

    func loginWithBiometric() async throws {
        let context = LAContext()
        let reason = "Log in to Camp Card"
        guard try await context.evaluatePolicy(.deviceOwnerAuthenticationWithBiometrics, localizedReason: reason) else {
            throw NetworkError.unauthorized
        }
        guard let refreshToken = keychain.biometricRefreshToken else {
            throw NetworkError.custom("Sign in with your password first to enable Face ID")
        }
        isLoading = true
        defer { isLoading = false }
        let resp: AuthResponse = try await api.request(.refresh(token: refreshToken))
        keychain.storeAuthTokens(access: resp.accessToken, refresh: resp.refreshToken)
        user = resp.user
        isAuthenticated = true
    }

    // MARK: - Logout
    func logout() {
        Task {
            try? await api.requestVoid(.logout)
        }
        keychain.clearAll()
        user = nil
        isAuthenticated = false
    }

    // MARK: - Update User
    func updateUser(_ updated: User) {
        user = updated
    }
}
